lib/src/main/java/io/ably/lib/transport/ConnectionManager.java

```diff
@@ -1193,19 +1194,19 @@ public void onMessage(ITransport transport, ProtocolMessage message) throws Ably
     }
 
     private void onChannelMessage(ProtocolMessage message) {
-        if(message.connectionSerial != null) {
-            connection.serial = message.connectionSerial.longValue();
-            if (connection.key != null)
-                connection.recoveryKey = connection.key + ":" + message.connectionSerial;
-        }
         channels.onMessage(message);
+        connection.recoveryKey = connection.createRecoveryKey();
     }
 
     private synchronized void onConnected(ProtocolMessage message) {
         final ErrorInfo error = message.error;
         boolean reattachOnResumeFailure = false; // this will indicate that channel must reattach when connected
         // event is received
 
+        boolean isConnectionResumeOrRecoverAttempt = !StringUtils.isNullOrEmpty(connection.key) ||
+            !StringUtils.isNullOrEmpty(ably.options.recover);
+        ably.options.recover = null; // RTN16k, explicitly setting null, so it won't be used for subsequent connection requests
+
         connection.reason = error;
         if (connection.id != null) { // there was a previous connection, so this is a resume and RTN15c applies
             Log.d(TAG, "There was a connection resume");
```

Comment:
Should we reset `msgSerial` here under `else` branch of `message.connectionId.equals(connection.id)` condition according to RTN16f

Reply:
yes https://github.com/ably/ably-java/commit/b42ff87ad98c55eeefb13d6f372e843fd5511cf2